File: Cpp/1254_number_of_closed_islands.cpp
// https://leetcode.cn/problems/number-of-closed-islands/

class Solution {
public:
    int closedIsland(vector<vector<int>>& grid) {
        vector<pair<int, int>> dirs = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        int m = grid.size(), n = grid[0].size();
        int res = 0;
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (grid[i][j] == 1) continue;
                queue<pair<int, int>> q;
                q.push(make_pair(i, j));
                grid[i][j] = 1;
                bool closed = true;
                while (!q.empty()) {
                    auto [curi, curj] = q.front();
                    q.pop();
                    if (curi == 0 || curi == m-1 || curj == 0 || curj == n-1) {
                        closed = false;
                    }
                    for (auto d: dirs) {
                        int x = curi + d.first, y = curj + d.second;
                        if (x >= 0 && x < m && y >= 0 && y < n && grid[x][y] == 0) {
                            q.push(make_pair(x, y));
                            grid[x][y] = 1;
                        }
                    }
                }
                if (closed) {
                    res += 1;
                }
            }
        }
        return res;
    }
};

File: Python3/1254_number_of_closed_islands.py
# https://leetcode.cn/problems/number-of-closed-islands/

class Solution:
    def closedIsland(self, grid):
        """
        Inputs:
            grid: list[list[int]]
        Outputs:
            res: int
        """
        m, n = len(grid), len(grid[0])
        res = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 0:
                    q = deque()
                    q.append((i, j))
                    grid[i][j] = 1
                    closed = True
                    while q:
                        curi, curj = q.popleft()
                        if curi == 0 or curi == m-1 or curj == 0 or curj == n-1:
                            closed = False
                        for x, y in [(curi-1, curj), (curi+1, curj), (curi, curj-1), (curi, curj+1)]:
                            if x >= 0 and x < m and y >= 0 and y < n and grid[x][y] == 0:
                                q.append((x, y))
                                grid[x][y] = 1
                    if closed:
                        res += 1
        return res